decisions/adr-0007-vulcan-workflows.md

```diff
@@ -0,0 +1,60 @@
+
+The second are data analysis pipelines (DAP), which operate on the research-ready data produced by the DPPs. They may or may not be compute-intensive, but are often run iteratively to refine analytic results or explore data. DAPs may be expected to produce images, tables and other artifacts as ways to introspect on data.
+
+Currently, DPPs are created and maintained outside of the library, and DAPs are created and maintained as part of the Vulcan code base. In addition, DPPs are configured and run manually by data scientists on external infrastructure, while DAPs run within Vulcan.
+
+While these two kinds of pipeline thus appear to be separate domains at the moment, there are also a variety of factors drawing them together. In both cases, the pipeline consists of a directed acyclic graph (DAG) of tasks; steps in the pipeline operate on files and yield files as output; each step ought to run in a container and must have the appropriate libs, packages, etc., available to it; the DAG must be run by a resource-aware scheduler that is able to queue work when constrained.
+
+In other words, in terms of execution, the two kinds of pipeline have virtually identical requirements; it cannot be assumed that they should therefore be executed in different computing environments. Both DPPs and DAPs are capable of generating system requirements that could benefit from access to HPC computing resources.
+
+## Decision
+
+We wish to unify the space of workflows in the data library, and integrate the normal work of data scientists in our group (and others) into the library. To do so, we will introduce a new Workflow data store in Vulcan. The data store will consist of two parts, a database model and a git tree. Workflow records will associate at the project level (i.e., each project will have a list of workflows). There will be no workflow author; the workflow belongs to the project.
+
+### Project workflows
+The Vulcan API will allow creating and listing project workflows. Each workflow can be also be browsed via the Vulcan UI. Workflows can still be run to produce Figures in Vulcan, which allow the workflow to be parameterized and run via the Vulcan UI. The Vulcan "steps" UI will be improved to monitor progress of a running workflow and show logging.
+
+The Vulcan API will also expose a git remote endpoint for each project workflow. When the workflow is created, it will be initialized to an empty tree or forked from an existing workflow. This is a normal git remote and can be pushed/pulled from as desired; the endpoint can possibly validate the push using an update hook.
+
+### Using Snakemake for workflows
+
+The format of a Vulcan workflow is a Snakemake workflow. This may include steps in python, R, etc. Workflows may define containers via a Snakefile and conda yaml. Vulcan will keep track of containers and re-use them across workflows.
```

Comment:
perhaps nit-level, and perhaps not the exact intent here, but conda recipes for R packages can lag behind what's released in CRAN and Bioconductor.  conda maintenance of R setups isn't as common as for python.

Reply:
Yeah, I'm fretting about this. I'd say R package management and containerization is "not good" in general. The conda yml has some advantages, but nothing precludes some other mechanism for defining containers being used and the resulting image referenced in the Snakefile.

Reply:
Definitely a fair read on R package menagement and containerization.  I think conda could work out okay, and with other languages in mind I agree it's the right place to start.  

The primary bit I'd like to be sure we can do, is grab most recent versions when I'm doing side-by-side workflow & dittoVerse dev.  This generally would extend past 'released' versions anyway, so an Rscript system that installes dittoViz & dittoSeq, when needed, from github could suffice.